docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+difference is that in the existing design, there is no way to parameterize
+provisioners or have more than one provisioner per cluster.  Therefore, once
+we know that we have to provision a volume to satisfy a claim, we need to
+determine which storage class to use for the claim.
+
+### How should a claim be matched to a storage class?
+
+We've discussed so far that volume claims and storage classes will be matched
+using the claim's label selector.  With that said, there are a number of small
+details to think about:
+
+1.  What should happen if there are multiple storage classes that match a
+    claim's selector?
+2.  Should we provide a way to specify a default storage class for claims
+    without a selector
+3.  Should we provide a way to specify a default storage class for claims
+    whose selector does not match any storage classes?
+
+For now, we will think of the problem of matching a claim to a storage class
+as a black box that solves the problem of: _given a claim and a list of
+storage classes that match that claim's selector, find a storage class to
+provision a volume for the claim with_.  Initially, we will implement the
+following behavior for this black box:
+
+1.  If multiple storage classes match the claim's selector, choose one at
+    random
+2.  If no storage class matches the claim's selector, use a default
+    class, which is specified as an argument to the controller manager
+
+### Controller workflow for provisioning volumes
+
+The workflow for provisioning does not need to change much to use storage
+classes.  The new workflow will be:
+
+1.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+2.  If no volume is found for the claim, the controller will attempt to
+    determine a storage class for the volume
+3.  If no storage class is found, the controller eventually retries finding
+    a storage class
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin`, which has a method called `NewProvisioner`
+    that returns a new provisioner.
+5.  The provisioner's `Provision` method has the same responsibility as it
+    does now, but it is now passed both the claim and storage class as
+    parameters
+6.  If `Provision` returns an error, the controller stops trying to
+    provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation for the `api.PersistentVolume` fails, it is
+      retried
+  2.  If the create operation never succeeds, the controller attempts to
+      delete the provisioned volume and creates an event on the claim
+
+## Proposed Design
+
+We propose that:
+
+1.  A new API group called `storage` be created to hold the a `StorageClass`
```

Comment:
ISTR some discussion with Brian, and wasn't he against a new Kind for this?  Or at least had some concerns?

The alternative is abusing ConfigMap, so...